[Tools] delete_candidate and delete_timepoint revamp

ridz1208 opened the pull request:
Original PR:  https://github.com/aces/Loris/pull/2425

Stella: Refactored the delete_candidate script to be independent from delete_timepoint.

me: Refactored delete_candidate to be using delete_session to avoid code duplication

kongtiaowang commented:
@ridz1208 
When I run "php delete_candidate.php delete_candidate 300040 MTL040 confirm".
I got this error.
"Dropping all DB entries for candidate CandID: 300040And PSCID: MTL040
PHP Warning:  PDO::exec(): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`wangshen_sandbox`.`session`, CONSTRAINT `FK_session_1` FOREIGN KEY (`CandID`) REFERENCES `candidate` (`CandID`)) in /var/www/loris/php/libraries/Database.class.inc on line 513
PHP Stack trace:
PHP   1. {main}() /var/www/loris/tools/delete_candidate.php:0
PHP   2. deleteCandidate() /var/www/loris/tools/delete_candidate.php:101
PHP   3. Database->delete() /var/www/loris/tools/delete_candidate.php:190
PHP   4. PDO->exec() /var/www/loris/php/libraries/Database.class.inc:513
Could not execute DELETE FROM candidate WHERE `CandID`='300040'. Stack trace#0 /var/www/loris/tools/delete_candidate.php(190): Database->delete('candidate', Array)
#1 /var/www/loris/tools/delete_candidate.php(101): deleteCandidate('300040', 'MTL040', true, false, Object(Database), '')
#2 {main}
PHP Fatal error:  Uncaught exception 'DatabaseException' with message 'Database DELETE did not execute successfully.Array
(
    [0] => 23000
    [1] => 1451
    [2] => Cannot delete or update a parent row: a foreign key constraint fails (`wangshen_sandbox`.`session`, CONSTRAINT `FK_session_1` FOREIGN KEY (`CandID`) REFERENCES `candidate` (`CandID`))
)
' in /var/www/loris/php/libraries/Database.class.inc:515
Stack trace:
#0 /var/www/loris/tools/delete_candidate.php(190): Database->delete('candidate', Array)
#1 /var/www/loris/tools/delete_candidate.php(101): deleteCandidate('300040', 'MTL040', true, false, Object(Database), '')
#2 {main}
  thrown in /var/www/loris/php/libraries/Database.class.inc on line 515
"

sruthymathew123 commented:
@ridz1208 
Test Summary(delete_candidate_script)
- Tried with all the three possibilities of the script.
- Ran php delete_candidate.php delete_candidate CandID PSCID for different data sets (like newly created candidate and with candidate containing so much data; all the data got fetched)
- Ran php delete_candidate.php delete_candidate CandID PSCID to sql and the delete queries are visible under my project/tables_sql
- ran php delete_candidate.php delete_candidate CandID PSCID confirm with a newly created candidate and all data related got deleted.

Test Summary (delete_timepoint_script)
- Did the similar thing with delete_timepoint_script also and everything was working fine.
- It will be more pleasing if you could add some meaningful messages like Dropping.../Generating...as used in delete_candidate_php in this script also.

Jkat commented:
Line 226 will fail,

```
            $DB->delete('feedback_bvl_entry', array('FeedbackID' => $id));
```
should be 
```
            $DB->delete('feedback_bvl_entry', array('FeedbackID' => $id['FeedbackID']));
```

kongtiaowang commented:
@ridz1208 It works fine for deleting candidate. It just can't delete this candidate. Maybe it is my test database cause the issue.

Jkat commented:
Still in `tools/delete_timepoint.php`:

Deletes from `feedback_bvl_thread` should happen after `feedback_bvl_entry` deletes.
Deletes from `session` should happen after `feedback_bvl_thread`.

This applies to both `if ($confirm)` and `elseif ($printToSQL)` sections

ridz1208 commented:
@Jkat give it a try

ridz1208 commented:
@kongtiaowang maybe try it now ?

ridz1208 commented:
@Jkat I think the ordering is correct feedback_bvl_entry and feedback_bvl_thread

ridz1208 commented:
@Jkat anymore inputs?

Jkat commented:
@ridz1208 For now no, I've only had time to look over the timepoint script

johnsaigle commented on tools/delete_candidate.php:
```diff
@@ -26,264 +41,192 @@
  * @link     https://www.github.com/aces/Loris-Trunk/
  */
 
+// Possible script actions
```
This section seems unnecessary now. The only valid action is delete_candidate. Also makes execution a little clunky e.g. "Example: php delete_candidate.php **delete_candidate** 965327 dcc0007".

I'd remove the command line parameter and switch-case statements.

ridz1208 replied:
@johnsaigle this has come up before, since I was just cleaning up stella's scripts we had decided to keep it as is but considering the amount of changes I've already done... I think this should be considered.

another thing to think about is that if someone is looking to expand this script then the infrastructure is already there, this is sort of an example on how to do it if you want

johnsaigle commented on tools/delete_candidate.php:
```diff
@@ -26,264 +41,192 @@
+    //find sessions
+    $sessions = $candidate->getListOfTimePoints();
+    if (is_null($sessions) || empty($sessions)) {
+        echo "There are no corresponding session for Candid : $CandID \n";
```
Change to "There are no corresponding sessions for CandID: $CandID"

johnsaigle commented on tools/delete_candidate.php:
```diff
@@ -26,264 +41,192 @@
+    fclose($fp);
+}
+
+if ($confirm === false && $printToSQL === false) {
```
add `&& !empty(sessions)`. Currently this message is a little strange when there are no sessions to be deleted and the script offers to delete the non-existent sessions.

ridz1208 replied:
@johnsaigle this would still create the delete of candidate parameters and other candidate specific fields (not related to timepoints)

johnsaigle commented:
Script is working well! 🌮 

Is it possible to clean up the printing a bit? The old script printed fairly cleanly but this one spits a lot of data to the screen and a lot of it is empty fields. Also, some prompts seem to be repeated.

I chose an arbitrary candidate on the Raisinbread test DB and got a lot of messy output as well as some repeated prompt messages. See the attached file.
[delete_candidate_output.txt](https://github.com/aces/Loris/files/876717/delete_candidate_output.txt)

ridz1208 commented:
@kongtiaowang I think this particular candidate has a problem in your database
maybe we can look at this together

ridz1208 commented:
@johnsaigle This version of the script uses the delete_timepoint scripts to avoid code duplication. I'm not sure if there is a way to not print out the output in patches, as for the things are are empty, those seem to be instrument data i can defenetly remove I dnt even know why it is there.

Those scripts are being used by admins only so usually having the most amount of info in the log is better even if there are empty arrays (which signifies that no data existed for this table for this candidate)

johnsaigle commented:
@ridz1208 Fair enough but maybe a -v verbose flag would be nice? But anyway we can save that for another iteration.

ridz1208 commented:
@johnsaigle I addressed as much as possible changes you have suggested, I modified slightly the output to make it more readable.

all data deleted by these scripts is not undoable and for this reason I am reluctant to remove all outputs as those print_r statements reflect all the data being deleted. I did however remove the instrument data as it overcrowds the output even on a very small number of instruments.

I also added some markers for timepoint deletion (####) and table deletions (-----)

Jkat commented on tools/delete_candidate.php:
```diff
@@ -26,264 +41,195 @@
+
+    // IF CONFIRMED, DELETE CANDIDATE
+    if ($confirm) {
+        echo "\nDropping all DB entries for candidate CandID: " . $CandID . "And PSCID: " .
```
Super duper incredibly minor detail, but a space could be added before the "And PSCID: " string, so: " And PSCID: " 
Same for line 195.

ridz1208 commented:
@Jkat done and done
